mm/pmm.c

```diff
@@ -302,12 +309,25 @@ static size_t process_memory_range(unsigned index, unsigned first_avail_region)
 
     /* Add initial 4K frames for early memory. */
     while (cur < MB(EARLY_VIRT_MEM) && cur + PAGE_SIZE <= end) {
-        add_early_frame(paddr_to_mfn(cur), PAGE_ORDER_4K);
+        // if the current region overlaps with the reserved region we skip it
+        if (!ranges_overlap(cur, ORDER_TO_SIZE(PAGE_ORDER_4K), reserved_start,
+                            reserved_size)) {
+            add_early_frame(paddr_to_mfn(cur), PAGE_ORDER_4K);
+        }
         cur += ORDER_TO_SIZE(PAGE_ORDER_4K);
     }
 
     while (cur + PAGE_SIZE <= end) {
         unsigned int order = PADDR_TO_ORDER(cur);
+
+        // make sure we don't overlap with the reserved region
+        while (ranges_overlap(cur, ORDER_TO_SIZE(order), reserved_start, reserved_size)) {
+            if (order == PAGE_ORDER_4K) {
+                goto skip;
+            }
+            --order;
+        }
+
```

Comment:
Maybe this check could be a part of the `add_early_frame()` and `add_frame()` so we always check if specified reserved ranges are not picked up? Just a loose suggestion to be evaluated given the other uses of `add_frame()`